### PR TITLE
Improve perfomance

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️

Related: #165.

### CODE REVIEW

File: src/Exception/DefaultRolesNotFoundException.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Rbac\Exception;
+
+use RuntimeException;
+use Yiisoft\FriendlyException\FriendlyExceptionInterface;
+
+class DefaultRolesNotFoundException extends RuntimeException implements FriendlyExceptionInterface
```
Comment (vjik) [inline src/Exception/DefaultRolesNotFoundException.php:10]: Add info about rename and finalize exception to changelog

File: src/ItemsStorageInterface.php [hunk 2]
```diff
@@ -7,7 +7,7 @@
 /**
  * A storage for RBAC roles and permissions used in {@see Manager}.
  *
- * @psalm-type ItemsIndexedByName = array<string, Permission|Role>
+ * @psalm-type ItemsIndexedByName = array<string, Item>
```
Comment (vjik) [inline src/ItemsStorageInterface.php:10]: Why? Storage keep permissions and roles, right?

File: src/Manager.php [hunk 3]
```diff
@@ -52,100 +50,65 @@ public function userHasPermission(
...
-        return $this->userHasItem(
-            $userId,
-            $this->itemsStorage->getPermission($permissionName),
-            $parameters,
-            $assignments
-        );
+        $parentPermissions = $this->itemsStorage->getParents($permission->getName());
+        $permissionNames = [$permission->getName(), ...array_keys($parentPermissions)];
+
+        return $this->assignmentsStorage->userHasItem($userId, $permissionNames);
```
Comment (vjik) [inline src/Manager.php:66]: Seems, this methods does not account that parent permissions also can contain rules, right?

File: src/ItemsStorageInterface.php (hunk 2 above)
Comment (arogachev) [reply src/ItemsStorageInterface.php:10]: `Item` is an abstract class and there are no more items other than permissions and roles. So basically it's equivalent form of writing, but a bit shorter.

File: src/Manager.php (hunk 3 above)
Comment (arogachev) [reply src/Manager.php:66]: I think this how it was before these changes too and suggest to solve it separately.

Comment (arogachev) [reply src/Manager.php:66]: #193

**Pull request merged**